### PR TITLE
applications: nrf5340_audio: Bugfix packet size check

### PR DESCRIPTION
**Pull request by Khuongh**
Added small bugfix to check if Rx size is the same as codec

OCT-2399

### CODE REVIEW

**Comment by Khuongh** [comment]
Managed to close the old PR.... So created a new one

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_headset.c [hunk 1]
```diff
@@ -677,14 +677,17 @@ int le_audio_user_defined_button_press(enum le_audio_user_defined_action action)
 	return 0;
 }
 
-int le_audio_config_get(uint32_t *bitrate, uint32_t *sampling_rate)
+int le_audio_config_get(uint32_t *bitrate, uint32_t *sampling_rate, uint32_t *octets_per_frame)
```
Comment (rick1082) [inline applications/nrf5340_audio/src/bluetooth/le_audio_cis_headset.c:680]: I think this PR great!
But I'm wondering if we can prevent adding new parameter in 'le_audio_config_get', but just calculate the octets from bitrate when we need. This is something we can discuss.

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_headset.c [hunk 2]
```diff
@@ -677,14 +677,17 @@ int le_audio_user_defined_button_press(enum le_audio_user_defined_action action)
...
 {
 	/* Get the configuration for the sink stream */
 	int frames_per_sec = 1000000 / bt_codec_cfg_get_frame_duration_us(audio_streams[0].codec);
-	int bits_per_frame = bt_codec_cfg_get_octets_per_frame(audio_streams[0].codec) * 8;
+	int octets_per_sdu = bt_codec_cfg_get_octets_per_frame(audio_streams[0].codec);
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/le_audio_cis_headset.c:684]:
```suggestion
	*octets_per_frame = bt_codec_cfg_get_octets_per_frame(audio_streams[0].codec);
```

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_headset.c [hunk 3]
```diff
@@ -677,14 +677,17 @@ int le_audio_user_defined_button_press(enum le_audio_user_defined_action action)
...
+	int bits_per_frame = octets_per_sdu * 8;
 
 	*sampling_rate = bt_codec_cfg_get_freq(audio_streams[0].codec);
 	*bitrate = frames_per_sec * bits_per_frame;
+	*octets_per_frame = octets_per_sdu;
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/le_audio_cis_headset.c:689]:
```suggestion

```

**Review by koffes — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_headset.c [hunk 4]
```diff
@@ -677,14 +677,16 @@ int le_audio_user_defined_button_press(enum le_audio_user_defined_action action)
...
 {
 	/* Get the configuration for the sink stream */
+
 	int frames_per_sec = 1000000 / bt_codec_cfg_get_frame_duration_us(audio_streams[0].codec);
-	int bits_per_frame = bt_codec_cfg_get_octets_per_frame(audio_streams[0].codec) * 8;
 
+	*octets_per_frame = bt_codec_cfg_get_octets_per_frame(audio_streams[0].codec);
 	*sampling_rate = bt_codec_cfg_get_freq(audio_streams[0].codec);
-	*bitrate = frames_per_sec * bits_per_frame;
+	*bitrate = frames_per_sec * (*sampling_rate) * 8;
```
[the remark this answers is not in the document]
Comment (erikrobstad) [reply applications/nrf5340_audio/src/bluetooth/le_audio_cis_headset.c:688]: See le_audio_bis_headset.c for reference

**Review by erikrobstad — CHANGES_REQUESTED** [review]

**Review by koffes — APPROVED** [review]

File: applications/nrf5340_audio/src/audio/streamctrl.c [hunk 5]
```diff
@@ -170,6 +170,24 @@ static void le_audio_rx_data_handler(uint8_t const *const p_data, size_t data_si
...
+	uint32_t bitrate, sampling_rate, octets_per_frame;
+	static uint32_t data_size_mismatch_cnt;
+
+	if (CONFIG_AUDIO_DEV == HEADSET) {
+		ret = le_audio_config_get(&bitrate, &sampling_rate, &octets_per_frame);
+		if (ret) {
+			LOG_ERR("Failed to get configs");
+		}
+
+		if (data_size != octets_per_frame) {
```
Comment (alexsven) [inline applications/nrf5340_audio/src/audio/streamctrl.c:182]: This should be checked directly in stream_recv_cb, no need to first get the octets per frame and then check when everything is already in the le_audio files

**Review by alexsven — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/audio/streamctrl.c (hunk 5 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl.c:182]: Good point. 
That will significantly reduce the amount of changed code

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_headset.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (Khuongh) [reply applications/nrf5340_audio/src/bluetooth/le_audio_cis_headset.c:688]: I dont believe so, in le_audio_bis_headset.c the values are retrieved from "active_stream.codec", which is an instance of "audio_codec_info". While in le_audio_cis_headset.c the "codec" is an instance of "bt_codec" and the values are not easily available.

File: applications/nrf5340_audio/src/audio/streamctrl.c [hunk 6]
```diff
@@ -170,6 +170,24 @@ static void le_audio_rx_data_handler(uint8_t const *const p_data, size_t data_si
...
+			data_size_mismatch_cnt++;
+			if ((data_size_mismatch_cnt % 500U) == 1U) {
```
[the remark this answers is not in the document]
Comment (Khuongh) [reply applications/nrf5340_audio/src/audio/streamctrl.c]: Removed "U"

File: applications/nrf5340_audio/src/audio/streamctrl.c (hunk 5 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/audio/streamctrl.c]: Pushed a new version that checks this in the le_audio files.

File: applications/nrf5340_audio/src/audio/streamctrl.c [hunk 7]
```diff
@@ -170,6 +170,24 @@ static void le_audio_rx_data_handler(uint8_t const *const p_data, size_t data_si
 	ret = data_fifo_pointer_first_vacant_get(&ble_fifo_rx, (void *)&iso_received, K_NO_WAIT);
 	ERR_CHK_MSG(ret, "Unable to get FIFO pointer");
 
+	uint32_t bitrate, sampling_rate, octets_per_frame;
+	static uint32_t data_size_mismatch_cnt;
+
+	if (CONFIG_AUDIO_DEV == HEADSET) {
+		ret = le_audio_config_get(&bitrate, &sampling_rate, &octets_per_frame);
+		if (ret) {
+			LOG_ERR("Failed to get configs");
```
[the remark this answers is not in the document]
Comment (Khuongh) [reply applications/nrf5340_audio/src/audio/streamctrl.c]: Removed from the new version

File: applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c [hunk 8]
```diff
@@ -175,6 +175,14 @@ static void stream_recv_cb(struct bt_audio_stream *stream, const struct bt_iso_r
 		bad_frame = true;
 	}
 
+	if (buf->len != active_stream.codec->octets_per_sdu) {
+		data_size_mismatch_cnt++;
+		if ((data_size_mismatch_cnt % 500) == 1) {
+			LOG_WRN("Data size mismatch, received: %d, codec: %d, total: %d", buf->len,
```
Comment (alexsven) [inline applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c:181]: Should be a LOG_DBG. Zero size packets are sent from time to time if there is no data to be sent on the gateway side. This goes for all three of them

**Review by alexsven — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c (hunk 8 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c]: Changed to LOG_DBG

File: applications/nrf5340_audio/src/audio/streamctrl.c [hunk 9]
```diff
@@ -170,6 +170,24 @@ static void le_audio_rx_data_handler(uint8_t const *const p_data, size_t data_si
 	ret = data_fifo_pointer_first_vacant_get(&ble_fifo_rx, (void *)&iso_received, K_NO_WAIT);
 	ERR_CHK_MSG(ret, "Unable to get FIFO pointer");
 
+	uint32_t bitrate, sampling_rate, octets_per_frame;
+	static uint32_t data_size_mismatch_cnt;
+
+	if (CONFIG_AUDIO_DEV == HEADSET) {
+		ret = le_audio_config_get(&bitrate, &sampling_rate, &octets_per_frame);
```
[the remark this answers is not in the document]
Comment (Khuongh) [reply applications/nrf5340_audio/src/audio/streamctrl.c]: New version does not use "le_audio_config_get"

File: applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c [hunk 10]
```diff
@@ -175,6 +175,14 @@ static void stream_recv_cb(struct bt_audio_stream *stream, const struct bt_iso_r
 		bad_frame = true;
 	}
 
+	if (buf->len != active_stream.codec->octets_per_sdu) {
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c:178]: To me it seems cleaner to do bt_codec_cfg_get_octets_per_frame(stream->codec) here. In other words use the stream variable inputted to this callback

**Review by erikrobstad — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c (hunk 10 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c]: Changed

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_headset.c [hunk 11]
```diff
@@ -441,6 +441,16 @@ static void stream_recv_cb(struct bt_audio_stream *stream, const struct bt_iso_r
 		bad_frame = true;
 	}
 
+	uint32_t octets_per_frame = bt_codec_cfg_get_octets_per_frame(audio_streams[0].codec);
```
[the remark this answers is not in the document]
Comment (Khuongh) [reply applications/nrf5340_audio/src/bluetooth/le_audio_cis_headset.c]: Changed

File: applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c [hunk 12]
```diff
@@ -413,6 +414,16 @@ static void stream_recv_cb(struct bt_audio_stream *stream, const struct bt_iso_r
 		return;
 	}
 
+	uint32_t octets_per_frame = bt_codec_cfg_get_octets_per_frame(stream->codec);
+
+	if (buf->len != octets_per_frame) {
+		data_size_mismatch_cnt++;
+		if ((data_size_mismatch_cnt % 500) == 1) {
+			LOG_DBG("Data size mismatch, received: %d, codec: %d, total: %d", buf->len,
+				octets_per_frame, data_size_mismatch_cnt);
```
[the remark this answers is not in the document]
Comment (Khuongh) [reply applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:582]: Added return, but will it be a problem if there's a mismatch and the log level is DBG and not warning?

**Review by erikrobstad — APPROVED** [review]
Looks good :)

**Comment by erikrobstad** [comment]
@koffes I think you have to resolve your old comments

**Pull request merged**